tiledb/sm/query/reader.cc

```diff
@@ -1636,13 +1637,40 @@ void Reader::fill_dense_coords_col_slab(
 
 Status Reader::unfilter_tiles(
     const std::string& name,
-    const std::vector<ResultTile*>& result_tiles) const {
+    const std::vector<ResultTile*>& result_tiles,
+    const std::vector<ResultCellSlab>* const result_cell_slabs) const {
   STATS_FUNC_IN(reader_unfilter_tiles);
 
   auto var_size = array_schema_->var_size(name);
   auto num_tiles = static_cast<uint64_t>(result_tiles.size());
   auto encryption_key = array_->encryption_key();
 
+  // Build an association from the result tile to the cell slab ranges
+  // that it contains.
+  std::unordered_map<
+      ResultTile*,
+      std::forward_list<std::pair<uint64_t, uint64_t>>>
+      cs_ranges;
+
+  // If 'result_cell_slabs' is NULL, the caller intends to bypass
+  // selective unfiltering.
+  if (result_cell_slabs) {
+    std::vector<ResultCellSlab>::const_reverse_iterator rit =
+        result_cell_slabs->rbegin();
+    while (rit != result_cell_slabs->rend()) {
```

Comment:
This will most probably be a perf straggler. In the very least, we need to parallelize. I see you are using a map and a forward list above, but we will have to be creative.